### PR TITLE
fix: API's categories indexation

### PR DESCRIPTION
**Pull request by jgiovaresco**
## Description

At startup, an initializer triggers an indexation for all Database APIs. Unfortunately, this initializer was not "adapting" API like the regular indexation happening when creating/updating API.

This PR brings a change to use the same adaptation logic for indexation for V4 API.
For V2, I only fixed something related to categories indexation.

## Additional context

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-fhonsfdald.chromatic.com)

### CODE REVIEW

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/inmemory/ApiCategoryQueryServiceInMemory.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    public Collection<String> findApiCategoryKeys(Api api) {
+        var categoriesIds = api.getCategories();
+        if (categoriesIds == null || categoriesIds.isEmpty()) {
+            return Collections.emptySet();
+        }
+
+        return storage
+            .stream()
+            .filter(category -> categoriesIds.contains(category.getId()))
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/inmemory/ApiCategoryQueryServiceInMemory.java]: Add the same logic here as in the service: the entry in categoriesIds can also contain category keys.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/api/domain_service/ApiIndexerDomainService.java [hunk 2]
```diff
@@ -20,29 +20,63 @@
...
+    }
+
+    /**
+     * Build an {@link IndexableApi} from an {@link Api}.
+     * @param context The indexation context.
+     * @param apiToIndex The API to index
+     * @param primaryOwner The primary owner of the API.
+     * @return The {@link IndexableApi} to index.
+     */
+    public IndexableApi toIndexableApi(Indexer.IndexationContext context, Api apiToIndex, PrimaryOwnerEntity primaryOwner) {
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/api/domain_service/ApiIndexerDomainService.java]: It doesn't look like the `context` is used in the method?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/search/SearchEngineServiceImpl.java [hunk 3]
```diff
@@ -134,6 +123,7 @@ public void index(ExecutionContext executionContext, Indexable source, boolean l
             content.setAction(ACTION_INDEX);
             content.setId(source.getId());
             content.setClazz(source.getClass().getName());
+            content.setClazz(source.getClass().getName());
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/search/SearchEngineServiceImpl.java]: Is this a typo?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/upgrade/initializer/SearchIndexInitializer.java [hunk 4]
```diff
@@ -194,15 +199,20 @@ private CompletableFuture<?> runApiIndexationAsync(ExecutorService executorServi
         }
         try {
             if (api.getDefinitionVersion() == DefinitionVersion.V4) {
-                indexable = apiMapper.toEntity(executionContext, api, primaryOwner, null, false);
+                indexable =
+                    apiIndexerDomainService.toIndexableApi(
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/upgrade/initializer/SearchIndexInitializer.java:203]: What about having the method `toIndexableApi` being able to handler v4 + non-v4 APIs? That way the logic for creating an indexable is in the domain service.

Comment (jgiovaresco) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/upgrade/initializer/SearchIndexInitializer.java:203]: I'm not confident enough to handle V2 API 🙈

**Review by jourdiw — APPROVED** [review]

**Pull request merged**